File: fabric/src/main/java/me/lukiiy/discordBridge/DiscordEvents.java
package me.lukiiy.discordBridge;

import me.lukiiy.discordBridge.api.CommandPlate;
import me.lukiiy.discordBridge.api.DiscordContext;
import me.lukiiy.discordBridge.api.MessageParts;
import me.lukiiy.uniStyle.UniParser;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel;
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent;
import net.dv8tion.jda.api.events.interaction.command.GenericCommandInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.thread.member.ThreadMemberJoinEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.minecraft.network.chat.Component;
import org.jetbrains.annotations.NotNull;

import java.time.Duration;
import java.time.OffsetDateTime;
import java.util.concurrent.atomic.AtomicReference;

public class DiscordEvents extends ListenerAdapter {
    @Override
    public void onMessageReceived(@NotNull MessageReceivedEvent e) {
        DiscordBridge instance = DiscordBridge.getInstance();
        AtomicReference<DiscordContext> context = new AtomicReference<>(instance.getContext());
        Config config = DiscordBridge.getInstance().getConfig();
        UniParser parser = UniParser.getDEFAULT();

        if (e.getChannel() != context.get().getChannel()) return;

        Member member = e.getMember();
        AtomicReference<Message> msg = new AtomicReference<>(e.getMessage());

        MessageParts parts = MessageParts.from(
                config.getOrDefault("messages.minecraft.format", "").replace("(user)", DiscordBridge.getInstance().miniSerializableName(member)),
                config.getOrDefault("messages.minecraft.prefix", ""),
                member,
                msg.get(),
                context.get().getBot(),
                config.getBoolean("discord.ignoreBots"),
                !config.getOrDefault("messages.minecraft.reply.default", "").isBlank(),
                config.getOrDefault("messages.minecraft.reply.default", ""),
                config.getBoolean("messages.minecraft.reply.ignoreBot")
        );

        if (parts == null) return;

        Component prefix = parts.getPrefix().isBlank() ? Component.empty() : Component.Serializer.fromJson(parser.serialize(parts.getPrefix()));
        Component content = DSerialMoj.reformatComponent(Component.Serializer.fromJson(parser.serialize(parts.getContent())));
        Component formatted = Component.empty().append(prefix).append(content).append(" ").append(DSerialMoj.joinWith(Component.literal(" "), DSerialMoj.listAttachments(msg.get())));

        DiscordBridge.getInstance().playerCast(formatted);
    }

    @Override
    public void onGenericCommandInteraction(@NotNull GenericCommandInteractionEvent e) {
        DiscordBridge instance = DiscordBridge.getInstance();

        CommandPlate cmd = instance.getContext().getCommand(e.getName());
        if (cmd == null) return;

        cmd.interaction(e);
        DiscordBridge.LOGGER.info("[Discord] {} issued server command: /{}", e.getMember(), e.getFullCommandName());
    }

    @Override
    public void onThreadMemberJoin(@NotNull ThreadMemberJoinEvent e) {
        ThreadChannel thread = e.getThread();
        if (!thread.isPublic() || thread.isArchived()) return;

        if (Duration.between(thread.getTimeCreated(), OffsetDateTime.now()).toMinutes() < 1) {
            DiscordBridge instance = DiscordBridge.getInstance();
            String msg = instance.getConfig().get("messages.minecraft.threadCreation");
            if (msg == null || msg.isBlank()) return;

            Component fMsg = Component.Serializer.fromJson(UniParser.getDEFAULT().serialize(msg.replace("(name)", thread.getName())));
            DiscordBridge.getInstance().playerCast(fMsg);
        }
    }

    @Override
    public void onGuildMemberJoin(@NotNull GuildMemberJoinEvent e) {
        TextChannel sysChannel = e.getGuild().getSystemChannel();
        DiscordBridge instance = DiscordBridge.getInstance();

        if (sysChannel != null && sysChannel == instance.getContext().getChannel()) {
            String msg = instance.getConfig().get("messages.minecraft.userJoin");
            if (msg == null || msg.isBlank()) return;

            Component fMsg = Component.Serializer.fromJson(UniParser.getDEFAULT().serialize(msg.replace("(name)", instance.miniSerializableName(e.getMember()))));
            DiscordBridge.getInstance().playerCast(fMsg);
        }
    }
}

File: minestom/src/main/kotlin/me/lukiiy/discordBridge/listeners/DiscordEvents.kt
package me.lukiiy.discordBridge.listeners

import me.lukiiy.discordBridge.Main
import me.lukiiy.discordBridge.api.MessageParts
import me.lukiiy.discordBridge.api.serialize.DSerialAdvnt
import me.lukiiy.discordBridge.api.serialize.DSerialAdvnt.MINI
import me.lukiiy.discordBridge.api.serialize.DSerialAdvnt.fromDiscord
import net.dv8tion.jda.api.events.interaction.command.GenericCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.events.thread.member.ThreadMemberJoinEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.JoinConfiguration
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import net.minestom.server.MinecraftServer
import net.minestom.server.entity.Player
import java.time.Duration
import java.time.OffsetDateTime
import kotlin.collections.forEach

class DiscordEvents : ListenerAdapter() {
    override fun onMessageReceived(e: MessageReceivedEvent) {
        val context = Main.context
        val config = Main.config

        if (e.channel !== context!!.channel) return

        val member = e.member
        val msg = e.message

        val parts = MessageParts.from(
            config.getOrDefault("messages.minecraft.format", "")!!.replace("(user)", Main.miniSerializableName(member!!)),
            config.getOrDefault("messages.minecraft.prefix", "")!!,
            member,
            e.message,
            context.bot,
            config.getBoolean("discord.ignoreBots"),
            !config.getOrDefault("messages.minecraft.reply.default", "")!!.isBlank(),
            config.getOrDefault("messages.minecraft.reply.default", ""),
            config.getBoolean("messages.minecraft.reply.ignoreBot")
        ) ?: return

        val prefix = if (parts.prefix.isBlank()) Component.empty() else MINI.deserialize(parts.prefix)
        val content = fromDiscord(parts.content)
        val formatted = Component.empty().append(prefix).append(content).appendSpace().append(Component.join(JoinConfiguration.spaces(), DSerialAdvnt.listAttachments(msg)))

        MinecraftServer.getConnectionManager().onlinePlayers.forEach { it.sendMessage(formatted) }
        println(PlainTextComponentSerializer.plainText().serialize(formatted))
    }

    override fun onGenericCommandInteraction(e: GenericCommandInteractionEvent) {
        val cmd = Main.context!!.getCommand(e.name) ?: return

        cmd.interaction(e)
        println(PlainTextComponentSerializer.plainText().serialize(MINI.deserialize("[Discord] " + Main.miniSerializableName(e.member!!) + " issued server command: /" + e.fullCommandName)))
    }

    override fun onThreadMemberJoin(e: ThreadMemberJoinEvent) {
        val thread = e.getThread()
        if (!thread.isPublic || thread.isArchived) return

        if (Duration.between(thread.timeCreated, OffsetDateTime.now()).toMinutes() < 1) {
            val msg = MINI.deserialize(Main.config.get("messages.minecraft.threadCreation")?.replace("(name)", thread.name) ?: "")
            if (msg == Component.empty()) return

            MinecraftServer.getConnectionManager().onlinePlayers.forEach { it.sendMessage(msg) }
            println(PlainTextComponentSerializer.plainText().serialize(msg))
        }
    }
}
